server/src/com/thoughtworks/go/config/ConfigCipherUpdater.java

```diff
@@ -0,0 +1,130 @@
+            byte[] oldCipher = FileUtils.readFileToByteArray(originalCipherFile);
+            new CipherProvider(systemEnvironment).resetCipher();
+
+            byte[] newCipher = FileUtils.readFileToByteArray(cipherFile);
+
+            if (new String(newCipher).equals(new String(oldCipher))) {
+                LOGGER.warn("Unable to generate a new safe cipher. Your cipher is unsafe.");
+                FileUtils.deleteQuietly(originalCipherFile);
+                FileUtils.deleteQuietly(originalConfigFile);
+                return;
+            }
+            Document document = new SAXBuilder().build(configFile);
+            List<String> encryptedAttributes = Arrays.asList("encryptedPassword", "encryptedManagerPassword");
+            List<String> encryptedNodes = Arrays.asList("encryptedValue");
+            for (String attributeName : encryptedAttributes) {
+                XPath xpathExpression = XPath.newInstance(String.format("//*[@%s]", attributeName));
+                List<Element> encryptedPasswordElements = xpathExpression.selectNodes(document);
+                for (Element element : encryptedPasswordElements) {
+                    Attribute encryptedPassword = element.getAttribute(attributeName);
+                    encryptedPassword.setValue(reEncryptUsingNewKey(oldCipher, newCipher, encryptedPassword.getValue()));
+                    LOGGER.debug("Replaced encrypted value at {}", element.toString());
+                }
+            }
+            for (String nodeName : encryptedNodes) {
+                XPath xpathExpression = XPath.newInstance(String.format("//%s", nodeName));
+                List<Element> encryptedNode = xpathExpression.selectNodes(document);
+                for (Element element : encryptedNode) {
+                    element.setText(reEncryptUsingNewKey(oldCipher, newCipher, element.getValue()));
+                    LOGGER.debug("Replaced encrypted value at {}", element.toString());
+                }
+            }
+            XmlUtils.writeXml(document, outputStream);
+            FileUtils.writeByteArrayToFile(configFile, outputStream.toByteArray());
+            LOGGER.info("Successfully re-encrypted config");
+        } catch (Exception e) {
+            LOGGER.error("Re-keying of cipher failed with error: [{}]", e.getMessage(), e);
+            if (originalCipherFile.exists()) {
+                try {
+                    FileUtils.copyFile(originalCipherFile, cipherFile);
+                } catch (IOException e1) {
+                    LOGGER.error("Could not replace the cipher file [{}] with original one [{}], please do so manually. Error: [{}]", cipherFile.getAbsolutePath(), originalCipherFile.getAbsolutePath(), e.getMessage(), e);
+                    bomb(e1);
+                }
+            }
+        }
+    }
+
+    private String reEncryptUsingNewKey(byte[] oldCipher, byte[] newCipher, String encryptedValue) throws InvalidCipherTextException {
+        GoCipher cipher = new GoCipher();
```

Comment:
Can make this a field, if needed. Ok to leave it in too. I thought that every call to `CipherProvider#primeKeyCache` reads the cipher file. But, it is `private static volatile`. So, it doesn't.